src/tcache.c

```diff
@@ -682,25 +686,126 @@ tcache_init(tsd_t *tsd, tcache_slow_t *tcache_slow, tcache_t *tcache,
 		    &cur_offset);
 	}
 	/*
-	 * For small size classes beyond tcache_maxclass (i.e. nhbins < NBINS),
-	 * their cache bins are initialized to a state to safely and efficiently
-	 * fail all fastpath alloc / free, so that no additional check around
-	 * nhbins is needed on fastpath.
+	 * For small size classes beyond thread_tcache_max (i.e.
+	 * thread_nhbins < NBINS), their cache bins are initialized to a state
+	 * to safely and efficiently fail all fastpath alloc / free, so that
+	 * no additional check around thread_nhbins is needed on fastpath.
 	 */
-	for (unsigned i = nhbins; i < SC_NBINS; i++) {
+	for (unsigned i = thread_nhbins; i < SC_NBINS; i++) {
 		/* Disabled small bins. */
 		cache_bin_t *cache_bin = &tcache->bins[i];
 		void *fake_stack = mem;
 		size_t fake_offset = 0;
 
 		cache_bin_init(cache_bin, &tcache_bin_info[i], fake_stack,
 		    &fake_offset);
-		assert(tcache_small_bin_disabled(i, cache_bin));
+		assert(tcache_small_bin_disabled(i, cache_bin, thread_nhbins));
 	}
 
 	cache_bin_postincrement(mem, &cur_offset);
-	/* Sanity check that the whole stack is used. */
-	assert(cur_offset == tcache_bin_alloc_size);
+	if (config_debug) {
+		/* Sanity check that the whole stack is used. */
+		size_t size, alignment;
+		cache_bin_info_compute_alloc(tcache_bin_info, thread_nhbins,
+		    &size, &alignment);
+		assert(cur_offset == size);
+	}
+}
+
+JEMALLOC_ALWAYS_INLINE void
+thread_tcache_max_and_nhbins_init(tsd_t *tsd) {
+	size_t thread_tcache_max = tsd_thread_tcache_max_get(tsd);
+	unsigned thread_nhbins = tsd_thread_nhbins_get(tsd);
+
+	 /*
+	 * Only during the tcache initialization, can thread_tcache_max
+	 * exceed TCACHE_MAXCLASS_LIMIT and can thread_nhbins exceed
+	 * TCACHE_NBINS_MAX. Then, they should be initialized to
+	 * global_do_not_change_tcache_maxclass and
+	 * global_do_not_change_nhbins respectively.
+	 */
+	if (thread_tcache_max > TCACHE_MAXCLASS_LIMIT) {
```

Comment:
Let's cap the values on the mallctl side, so that the invalids ones never make into tsd.

Reply:
This is still needed because initializer needs a const value and the initial value set is `TCACHE_MAXCLASS_LIMIT + 1`. When encountering this, we know the tcache is not initialized yet and assign the real initial value we want, which is `global_do_not_change_tcache_maxclass`.

Reply:
We should be able to initialize them at the beginning of `tsd_tcache_enabled_data_init`:

https://github.com/jemalloc/jemalloc/blob/dev/src/tcache.c#L637

Then every access should be based on valid values.